consensus/service/src/byzantine_ledger/mod.rs

```diff
@@ -0,0 +1,617 @@
+// Copyright (c) 2018-2020 MobileCoin Inc.
+
+//! A Federated, Byzantine Fault-Tolerant Ledger.
+//!
+//! Orchestrates running single-slot consensus, or performing ledger sync with peers.
+
+mod ledger_sync_state;
+mod task_message;
+mod worker;
+
+use crate::{
+    byzantine_ledger::{task_message::TaskMessage, worker::ByzantineLedgerWorker},
+    counters,
+    tx_manager::TxManager,
+};
+use mc_common::{logger::Logger, NodeID, ResponderId};
+use mc_connection::{BlockchainConnection, ConnectionManager};
+use mc_consensus_scp::{scp_log::LoggingScpNode, Msg, Node, QuorumSet, ScpNode};
+use mc_crypto_keys::Ed25519Pair;
+use mc_ledger_db::Ledger;
+use mc_peers::{ConsensusConnection, ConsensusMsg, ThreadedBroadcaster, VerifiedConsensusMsg};
+use mc_transaction_core::tx::TxHash;
+use mc_util_metered_channel::{self, Sender};
+use std::{
+    path::PathBuf,
+    sync::{
+        atomic::{AtomicBool, AtomicU64, Ordering},
+        Arc, Mutex,
+    },
+    thread::{self, JoinHandle},
+    time::{Duration, Instant},
+};
+
+/// Time we're allowed to stay behind before we initiate catchup.
+/// This reduces the amount of unnecessary catchups due to minor network issues.
+const IS_BEHIND_GRACE_PERIOD: Duration = Duration::from_secs(10);
+
+/// Maximum number of pending values to hand over to `scp` at each slot.
+/// This is currently capped due to pending values not being capped and an outstanding issue of
+/// `scp` performing more expensive and exhaustive validation than is sometimes required.
+const MAX_PENDING_VALUES_TO_NOMINATE: usize = 100;
+
+pub struct ByzantineLedger {
+    sender: Sender<TaskMessage>,
+    thread_handle: Option<JoinHandle<()>>,
+    is_behind: Arc<AtomicBool>,
+    highest_peer_block: Arc<AtomicU64>,
+    highest_outgoing_consensus_msg: Arc<Mutex<Option<ConsensusMsg>>>,
+}
+
+impl ByzantineLedger {
+    pub fn new<
+        PC: BlockchainConnection + ConsensusConnection + 'static,
+        L: Ledger + Sync + 'static,
+    >(
+        node_id: NodeID,
+        quorum_set: QuorumSet,
+        peer_manager: ConnectionManager<PC>,
+        ledger: L,
+        tx_manager: Arc<Mutex<Box<dyn TxManager>>>,
```

Comment:
So, I'm surprised to still see the `Arc<Mutex>` being used on the `TxManager` given that that wasn't necessary before this revision and it caused performance problems before.

Normally to accomplish dependency injection, you only need `Box<dyn TxManager>`.
If we end up with extra mutexes that we didn't need before, that seems like a bad thing that will make it harder to reason about the system and its performance. And most of the function on the TxManager are thread-safe, so this is a significant pessimization.

You could try something with RwLock if you really need this?

Can you speak to why `Arc<Mutex>` is needed here?

Reply:
@garbageslam This file has conflicts with master, and the Arc<Mutex> should go away when that is resolved. It may actually be less work to close this PR and open a new one that starts from the current master, so I might do that instead.